File: project-cubie-a5e/board/radxa/cubie_a5e/rootfs-overlay/usr/bin/ping_rpmsg.py
#!/usr/bin/env python3
"""
ping_rpmsg.py - Standard Linux RPMsg Ping-Pong Host Benchmark (Python)

Target: Linux Host (ARM64 / x86_64) communicating with Allwinner T527 XuanTie E907
Protocol: Standard Linux virtio_rpmsg_bus character interface (/dev/rpmsg0 or /dev/rpmsg_ctrl0)

Features:
 - Event-driven I/O using select() with 0% idle CPU utilization
 - Automatic endpoint creation via /dev/rpmsg_ctrl0 if needed
 - Precise Round-Trip Time (RTT) measurements with nanosecond timestamps
 - Statistics: Min, Average, Max, Jitter, Percentiles (p50, p90, p99), Throughput
"""

import os
import sys
import time
import struct
import fcntl
import select
import argparse
from typing import List, Optional

# ANSI Escape Colors
C_RESET   = "\033[0m"
C_BOLD    = "\033[1m"
C_RED     = "\033[31m"
C_GREEN   = "\033[32m"
C_YELLOW  = "\033[33m"
C_BLUE    = "\033[34m"
C_CYAN    = "\033[36m"

# Standard Linux RPMsg ioctl definitions
# struct rpmsg_endpoint_info: char name[32], u32 src, u32 dst
RPMSG_CREATE_EPT_IOCTL = 0x4028B501   # _IOW(0xb5, 0x1, struct rpmsg_endpoint_info)
RPMSG_PING_EPT_ADDR    = 1024
RPMSG_PAYLOAD_FMT      = "<IQ48s"      # uint32 seq, uint64 ts_ns, char text[48]
RPMSG_PAYLOAD_LEN      = struct.calcsize(RPMSG_PAYLOAD_FMT)

def create_endpoint_from_ctrl(ctrl_path: str, name: str = "rpmsg-ping-channel",
                              src: int = RPMSG_PING_EPT_ADDR,
                              dst: int = RPMSG_PING_EPT_ADDR) -> Optional[str]:
    """Create an RPMsg endpoint via control device ioctl."""
    try:
        ctrl_fd = os.open(ctrl_path, os.O_RDWR)
    except OSError as e:
        print(f"{C_RED}[ERROR] Failed to open control device {ctrl_path}: {e}{C_RESET}")
        return None

    try:
        name_bytes = name.encode('utf-8')[:31]
        name_bytes = name_bytes.ljust(32, b'\x00')
        ept_info = struct.pack("32sII", name_bytes, src, dst)
        fcntl.ioctl(ctrl_fd, RPMSG_CREATE_EPT_IOCTL, ept_info)
        print(f"{C_GREEN}[INFO] Created endpoint '{name}' via {ctrl_path}{C_RESET}")
    except OSError as e:
        # If already exists, proceed
        print(f"{C_YELLOW}[WARN] ioctl CREATE_EPT returned: {e} (might already exist){C_RESET}")
    finally:
        os.close(ctrl_fd)

    # Wait for dynamic udev device creation
    candidate = "/dev/rpmsg0"
    for _ in range(10):
        if os.path.exists(candidate):
            return candidate
        time.sleep(0.05)
    return candidate

def auto_find_device() -> str:
    """Auto-detect available RPMsg devices in sysfs / dev."""
    candidates = [
        "/dev/rpmsg0",
        "/dev/rpmsg_ctrl0",
        "/dev/ttyRPMSG0",
    ]
    for c in candidates:
        if os.path.exists(c):
            return c
    return "/dev/rpmsg0"

def main():
    parser = argparse.ArgumentParser(description="Standard Linux RPMsg Ping-Pong Host Benchmark (Python)")
    parser.add_argument("-d", "--dev", default="", help="RPMsg device path (default: auto-detect)")
    parser.add_argument("-n", "--count", type=int, default=1000, help="Number of pings (default: 1000, 0=continuous)")
    parser.add_argument("-T", "--duration", type=float, default=0.0, help="Run duration in seconds (e.g. 60 for 1 min; overrides -n if > 0)")
    parser.add_argument("-s", "--sleep", type=int, default=1000, help="Sleep between pings in microseconds (default: 1000)")
    parser.add_argument("-p", "--payload", default="Ping from Linux RPMsg Python", help="Custom payload string")
    parser.add_argument("-t", "--timeout", type=float, default=1000.0, help="Pong timeout in milliseconds (default: 1000)")
    args = parser.parse_args()

    print(f"{C_CYAN}{C_BOLD}================================================================{C_RESET}")
    print(f"{C_CYAN}{C_BOLD}  Allwinner T527 Linux VirtIO RPMsg Ping-Pong Benchmark (Python){C_RESET}")
    print(f"{C_CYAN}  Protocol: Linux kernel virtio_rpmsg_bus (/dev/rpmsg0)         {C_RESET}")
    print(f"{C_CYAN}  Channel : rpmsg-ping-channel (Endpoint Addr: 1024)            {C_RESET}")
    if args.duration > 0:
        print(f"{C_CYAN}  Mode    : Timed run for {args.duration:.1f} seconds (live rate display) {C_RESET}")
    print(f"{C_CYAN}{C_BOLD}================================================================{C_RESET}\n")

    dev_path = args.dev if args.dev else auto_find_device()
    print(f"[INFO] Initial device selection: {dev_path}")

    # Handle control device endpoint creation if needed
    if "rpmsg_ctrl" in dev_path:
        created_dev = create_endpoint_from_ctrl(dev_path)
        if created_dev and os.path.exists(created_dev):
            dev_path = created_dev
        else:
            dev_path = "/dev/rpmsg0"

    print(f"{C_GREEN}[INFO] Opening RPMsg device: {dev_path}{C_RESET}")
    try:
        fd = os.open(dev_path, os.O_RDWR | os.O_NONBLOCK)
    except OSError as e:
        print(f"{C_RED}[FATAL] Could not open {dev_path}: {e}{C_RESET}")
        print(f"{C_YELLOW}[HINT] Ensure remoteproc is started and /sys/class/remoteproc/remoteproc0/state is 'running'.{C_RESET}")
        sys.exit(1)

    payload_bytes = args.payload.encode('utf-8')[:47].ljust(48, b'\x00')
    timeout_sec = args.timeout / 1000.0
    sleep_sec = args.sleep / 1_000_000.0

    latencies_us: List[float] = []
    seq = 0
    timeouts = 0
    total_tx_bytes = 0
    total_rx_bytes = 0
    bench_start_ns = time.monotonic_ns()
    last_report_ns = bench_start_ns
    last_report_seq = 0
    last_report_tx_bytes = 0
    last_report_rx_bytes = 0

    mode_str = f"{args.duration:.1f}s duration" if args.duration > 0 else f"{args.count} pings"
    print(f"[INFO] Starting benchmark ({mode_str}, delay={args.sleep}us, timeout={args.timeout}ms)...\n")

    try:
        while True:
            now_ns = time.monotonic_ns()
            elapsed_sec = (now_ns - bench_start_ns) / 1_000_000_000.0

            if args.duration > 0:
                if elapsed_sec >= args.duration:
                    break
            elif args.count > 0 and seq >= args.count:
                break

            seq += 1
            tx_ns = time.monotonic_ns()
            tx_pkt = struct.pack(RPMSG_PAYLOAD_FMT, seq, tx_ns, payload_bytes)

            try:
                os.write(fd, tx_pkt)
                total_tx_bytes += len(tx_pkt)
            except OSError as e:
                print(f"{C_RED}[ERROR] write() failed on seq={seq}: {e}{C_RESET}")
                break

            # Event-driven wait with select (0% CPU idle consumption)
            rlist, _, _ = select.select([fd], [], [], timeout_sec)
            rx_ns = time.monotonic_ns()

            if rlist:
                try:
                    rx_data = os.read(fd, 512)
                    if rx_data:
                        total_rx_bytes += len(rx_data)
                        rtt_us = (rx_ns - tx_ns) / 1000.0
                        latencies_us.append(rtt_us)
                    else:
                        timeouts += 1
                except OSError as e:
                    print(f"{C_YELLOW}[WARN] read() error on seq={seq}: {e}{C_RESET}")
                    timeouts += 1
            else:
                timeouts += 1
                if timeouts <= 5 or timeouts % 100 == 0:
                    print(f"{C_YELLOW}[WARN] Ping seq={seq} TIMEOUT ({args.timeout} ms){C_RESET}")

            if sleep_sec > 0:
                time.sleep(sleep_sec)

            # Live 1-second rolling rate & latency display
            if (now_ns - last_report_ns) >= 1_000_000_000:
                window_dt = (now_ns - last_report_ns) / 1_000_000_000.0
                window_pings = seq - last_report_seq
                rate = window_pings / window_dt
                window_bytes = (total_tx_bytes - last_report_tx_bytes) + (total_rx_bytes - last_report_rx_bytes)
                bw_kb = (window_bytes / window_dt) / 1024.0
                bw_str = f"{bw_kb/1024.0:5.2f} MB/s" if bw_kb >= 1024.0 else f"{bw_kb:5.1f} KB/s"
                last_rtt = latencies_us[-1] if latencies_us else 0.0
                recent_avg = sum(latencies_us[-window_pings:]) / max(1, min(window_pings, len(latencies_us))) if latencies_us else 0.0

                time_str = f"[{elapsed_sec:04.1f}s / {args.duration:04.1f}s]" if args.duration > 0 else f"[{elapsed_sec:04.1f}s]"
                sys.stdout.write(
                    f"\r  {C_BOLD}{C_GREEN}{time_str}{C_RESET} Rate: {C_BOLD}{rate:6.1f} pkts/s{C_RESET} | "
                    f"Speed: {C_BOLD}{C_CYAN}{bw_str}{C_RESET} | Total: {seq:,} | "
                    f"Last RTT: {last_rtt:6.1f} us | Avg RTT: {recent_avg:6.1f} us | "
                    f"Timeouts: {timeouts}  "
                )
                sys.stdout.flush()
                last_report_ns = now_ns
                last_report_seq = seq
                last_report_tx_bytes = total_tx_bytes
                last_report_rx_bytes = total_rx_bytes

    except KeyboardInterrupt:
        print(f"\n{C_YELLOW}[INFO] Interrupted by user.{C_RESET}")
    finally:
        os.close(fd)

    bench_end_ns = time.monotonic_ns()
    total_elapsed_ms = (bench_end_ns - bench_start_ns) / 1_000_000.0

    # Print Summary Statistics
    print(f"\n\n{C_CYAN}{C_BOLD}================================================================{C_RESET}")
    print(f"{C_CYAN}{C_BOLD}                    RPMsg Benchmark Results                     {C_RESET}")
    print(f"{C_CYAN}{C_BOLD}================================================================{C_RESET}")
    print(f"  Total Packets Sent : {C_BOLD}{seq}{C_RESET}")
    print(f"  Successful Replies : {C_GREEN}{len(latencies_us)}{C_RESET}")
    print(f"  Timed Out Packets  : {C_RED if timeouts else C_GREEN}{timeouts}{C_RESET}")
    print(f"  Total Test Time    : {total_elapsed_ms:.2f} ms")

    if latencies_us:
        min_lat = min(latencies_us)
        max_lat = max(latencies_us)
        avg_lat = sum(latencies_us) / len(latencies_us)

        # Variance / Jitter
        variance = sum((x - avg_lat) ** 2 for x in latencies_us) / len(latencies_us)
        jitter = variance ** 0.5

        sorted_lat = sorted(latencies_us)
        p50 = sorted_lat[int(len(sorted_lat) * 0.50)]
        p90 = sorted_lat[int(len(sorted_lat) * 0.90)]
        p99 = sorted_lat[int(len(sorted_lat) * 0.99)]
        throughput = (len(latencies_us) / (total_elapsed_ms / 1000.0)) if total_elapsed_ms > 0 else 0

        total_bytes = total_tx_bytes + total_rx_bytes
        bw_kb_sec = (total_bytes / (total_elapsed_ms / 1000.0)) / 1024.0 if total_elapsed_ms > 0 else 0.0
        bw_summary_str = f"{bw_kb_sec/1024.0:.2f} MB/s ({bw_kb_sec:.1f} KB/s)" if bw_kb_sec >= 1024.0 else f"{bw_kb_sec:.1f} KB/s"
        tx_kb_sec = (total_tx_bytes / (total_elapsed_ms / 1000.0)) / 1024.0 if total_elapsed_ms > 0 else 0.0
        rx_kb_sec = (total_rx_bytes / (total_elapsed_ms / 1000.0)) / 1024.0 if total_elapsed_ms > 0 else 0.0

        print(f"  Packet Rate        : {C_BOLD}{throughput:.1f} msgs/sec{C_RESET}")
        print(f"  Bandwidth (Total)  : {C_BOLD}{C_CYAN}{bw_summary_str}{C_RESET} (bidirectional)")
        print(f"  Bandwidth (TX)     : {tx_kb_sec:.1f} KB/s ({total_tx_bytes:,} bytes)")
        print(f"  Bandwidth (RX)     : {rx_kb_sec:.1f} KB/s ({total_rx_bytes:,} bytes)")
        print(f"----------------------------------------------------------------")
        print(f"  Min Latency        : {C_GREEN}{min_lat:.2f} us{C_RESET}")
        print(f"  Average Latency    : {C_BOLD}{avg_lat:.2f} us{C_RESET}")
        print(f"  Max Latency        : {C_YELLOW if max_lat < 500 else C_RED}{max_lat:.2f} us{C_RESET}")
        print(f"  Jitter (Std Dev)   : {jitter:.2f} us")
        print(f"----------------------------------------------------------------")
        print(f"  Percentile p50     : {p50:.2f} us")
        print(f"  Percentile p90     : {p90:.2f} us")
        print(f"  Percentile p99     : {p99:.2f} us")
    else:
        print(f"{C_RED}[FAIL] No pong replies received! All packets timed out.{C_RESET}")
        sys.exit(1)
    print(f"{C_CYAN}================================================================{C_RESET}\n")

if __name__ == "__main__":
    main()

File: riscv-firmware/apps/testPingRpmsg/linux/ping_rpmsg.cpp
/*
 * ping_rpmsg.cpp - Standard Linux RPMsg Ping-Pong Host Benchmark
 *
 * Target: Linux Host (ARM64 / x86_64) communicating with Allwinner T527 XuanTie E907
 * Protocol: Standard Linux virtio_rpmsg_bus character interface (/dev/rpmsg0 or /dev/rpmsg_ctrl0)
 *
 * Measures:
 *  - Round-Trip Time (RTT) per packet with nanosecond precision
 *  - Min, Average, Max latency and Jitter (Standard Deviation)
 *  - Latency Histogram / Percentiles (p50, p90, p99, p99.9)
 *  - Transmission throughput (msgs/sec)
 */

#include <iostream>
#include <iomanip>
#include <vector>
#include <algorithm>
#include <cmath>
#include <cstring>
#include <cstdint>
#include <cerrno>
#include <unistd.h>
#include <fcntl.h>
#include <sys/ioctl.h>
#include <sys/poll.h>
#include <time.h>
#include <getopt.h>

#if __has_include(<linux/rpmsg.h>)
#include <linux/rpmsg.h>
#else
struct rpmsg_endpoint_info {
    char name[32];
    uint32_t src;
    uint32_t dst;
};
#define RPMSG_CREATE_EPT_IOCTL  _IOW(0xb5, 0x1, struct rpmsg_endpoint_info)
#define RPMSG_DESTROY_EPT_IOCTL _IO(0xb5, 0x2)
#endif

#include "../../common/include/resource_table.h"

struct RpmsgPingPayload {
    uint32_t seq;
    uint64_t host_tx_ts_ns;
    char     text[48];
} __attribute__((packed));

static inline uint64_t get_time_ns() {
    struct timespec ts;
    clock_gettime(CLOCK_MONOTONIC_RAW, &ts);
    return (uint64_t)ts.tv_sec * 1000000000ULL + (uint64_t)ts.tv_nsec;
}

static void print_usage(const char *prog) {
    std::cout << "Usage: " << prog << " [options]\n"
              << "Options:\n"
              << "  -d, --dev <path>      RPMsg character device (default: /dev/rpmsg0 or /dev/rpmsg_ctrl0)\n"
              << "  -n, --count <num>     Number of pings to send (default: 1000, 0 = continuous)\n"
              << "  -s, --sleep <us>      Sleep between pings in microseconds (default: 1000)\n"
              << "  -p, --payload <str>   Custom payload string (max 47 chars)\n"
              << "  -t, --timeout <ms>    Pong timeout in milliseconds (default: 1000)\n"
              << "  -h, --help            Show this help message\n";
}

int main(int argc, char *argv[]) {
    uint32_t count = 1000;
    uint32_t sleep_us = 1000;
    uint32_t timeout_ms = 1000;
    std::string payload_str = "Ping from Linux RPMsg Host";
    std::string dev_path = "";

    static struct option long_options[] = {
        {"dev",     required_argument, 0, 'd'},
        {"count",   required_argument, 0, 'n'},
        {"sleep",   required_argument, 0, 's'},
        {"payload", required_argument, 0, 'p'},
        {"timeout", required_argument, 0, 't'},
        {"help",    no_argument,       0, 'h'},
        {0, 0, 0, 0}
    };

    int opt;
    while ((opt = getopt_long(argc, argv, "d:n:s:p:t:h", long_options, nullptr)) != -1) {
        switch (opt) {
            case 'd': dev_path = optarg; break;
            case 'n': count = std::stoul(optarg); break;
            case 's': sleep_us = std::stoul(optarg); break;
            case 'p': payload_str = optarg; break;
            case 't': timeout_ms = std::stoul(optarg); break;
            case 'h': print_usage(argv[0]); return 0;
            default:  print_usage(argv[0]); return 1;
        }
    }

    std::cout << "================================================================\n";
    std::cout << "  Allwinner T527 Linux VirtIO RPMsg Ping-Pong Benchmark        \n";
    std::cout << "  Protocol: Linux kernel virtio_rpmsg_bus                      \n";
    std::cout << "  Channel : rpmsg-ping-channel (Endpoint Addr: 1024)           \n";
    std::cout << "================================================================\n";

    // Auto-discover RPMsg device if not provided
    if (dev_path.empty()) {
        const char *candidate_devs[] = {
            "/dev/rpmsg0",
            "/dev/rpmsg_ctrl0",
            "/dev/ttyRPMSG0"
        };
        for (const char *cand : candidate_devs) {
            if (access(cand, F_OK) == 0) {
                dev_path = cand;
                break;
            }
        }
        if (dev_path.empty()) {
            dev_path = "/dev/rpmsg0"; // Default fallback
        }
    }

    std::cout << "[INFO] Target device: " << dev_path << "\n";

    int fd = -1;

    // Check if device is rpmsg_ctrlX and create an endpoint
    if (dev_path.find("rpmsg_ctrl") != std::string::npos) {
        std::cout << "[INFO] Detected RPMsg control device. Creating endpoint 'rpmsg-ping-channel'...\n";
        int ctrl_fd = open(dev_path.c_str(), O_RDWR);
        if (ctrl_fd < 0) {
            std::cerr << "[ERROR] Failed to open " << dev_path << ": " << strerror(errno) << "\n";
            return 1;
        }

        struct rpmsg_endpoint_info ept_info;
        memset(&ept_info, 0, sizeof(ept_info));
        strncpy(ept_info.name, "rpmsg-ping-channel", sizeof(ept_info.name) - 1);
        ept_info.src = RPMSG_PING_EPT_ADDR;
        ept_info.dst = RPMSG_PING_EPT_ADDR;

        if (ioctl(ctrl_fd, RPMSG_CREATE_EPT_IOCTL, &ept_info) < 0) {
            std::cerr << "[ERROR] RPMSG_CREATE_EPT_IOCTL failed: " << strerror(errno) << "\n";
            close(ctrl_fd);
            return 1;
        }
        close(ctrl_fd);
        dev_path = "/dev/rpmsg0";
        usleep(100000); // Allow udev to instantiate /dev/rpmsg0
    }

    fd = open(dev_path.c_str(), O_RDWR | O_NONBLOCK);
    if (fd < 0) {
        std::cerr << "[ERROR] Failed to open " << dev_path << ": " << strerror(errno) << "\n";
        std::cerr << "[HINT] Ensure Linux remoteproc firmware is booted and rpmsg_char driver is loaded.\n";
        std::cerr << "[HINT] modprobe rpmsg_char && modprobe virtio_rpmsg_bus\n";
        return 1;
    }

    std::cout << "[INFO] Successfully opened " << dev_path << ". Starting ping-pong loop...\n\n";

    std::vector<double> latencies_us;
    latencies_us.reserve(count > 0 ? count : 10000);

    uint64_t bench_start_ns = get_time_ns();
    uint32_t seq = 0;
    uint32_t timeouts = 0;

    for (uint32_t i = 0; count == 0 || i < count; ++i) {
        seq++;

        RpmsgPingPayload tx_payload;
        tx_payload.seq = seq;
        strncpy(tx_payload.text, payload_str.c_str(), sizeof(tx_payload.text) - 1);
        tx_payload.text[sizeof(tx_payload.text) - 1] = '\0';

        uint64_t tx_ns = get_time_ns();
        tx_payload.host_tx_ts_ns = tx_ns;

        // Send RPMsg Ping
        ssize_t bytes_written = write(fd, &tx_payload, sizeof(tx_payload));
        if (bytes_written < 0) {
            std::cerr << "[ERROR] write() failed on seq=" << seq << ": " << strerror(errno) << "\n";
            break;
        }

        // Wait for Pong Response using poll()
        struct pollfd pfd;
        pfd.fd = fd;
        pfd.events = POLLIN;

        int ret = poll(&pfd, 1, timeout_ms);
        if (ret > 0 && (pfd.revents & POLLIN)) {
            RpmsgPingPayload rx_payload;
            ssize_t bytes_read = read(fd, &rx_payload, sizeof(rx_payload));
            uint64_t rx_ns = get_time_ns();

            if (bytes_read > 0) {
                double rtt_us = (double)(rx_ns - tx_ns) / 1000.0;
                latencies_us.push_back(rtt_us);
            } else {
                timeouts++;
                std::cerr << "[WARN] Ping seq=" << seq << " read() error: " << strerror(errno) << "\n";
            }
        } else if (ret == 0) {
            timeouts++;
            std::cerr << "[WARN] Ping seq=" << seq << " TIMEOUT (" << timeout_ms << " ms)\n";
        } else {
            std::cerr << "[ERROR] poll() error on seq=" << seq << ": " << strerror(errno) << "\n";
            break;
        }

        if (sleep_us > 0) {
            usleep(sleep_us);
        }

        if (seq % 200 == 0) {
            std::cout << "  Progress: " << seq << " packets sent, last RTT: "
                      << std::fixed << std::setprecision(2) << latencies_us.back() << " us\r" << std::flush;
        }
    }

    uint64_t bench_end_ns = get_time_ns();
    double total_time_sec = (double)(bench_end_ns - bench_start_ns) / 1e9;

    std::cout << "\n\n==================== BENCHMARK RESULTS ====================\n";
    std::cout << "Packets Sent   : " << seq << "\n";
    std::cout << "Packets Recv   : " << latencies_us.size() << " ("
              << (seq > 0 ? (double)latencies_us.size() * 100.0 / seq : 0.0) << "% success)\n";
    std::cout << "Timeouts       : " << timeouts << "\n";
    std::cout << "Total Duration : " << std::fixed << std::setprecision(3) << total_time_sec << " s\n";
    std::cout << "Throughput     : " << std::fixed << std::setprecision(1)
              << ((double)latencies_us.size() / total_time_sec) << " msgs/sec\n";

    if (!latencies_us.empty()) {
        std::sort(latencies_us.begin(), latencies_us.end());

        double min_lat = latencies_us.front();
        double max_lat = latencies_us.back();
        double sum = 0;
        for (double l : latencies_us) sum += l;
        double avg_lat = sum / latencies_us.size();

        double variance = 0;
        for (double l : latencies_us) {
            variance += (l - avg_lat) * (l - avg_lat);
        }
        double jitter_stddev = std::sqrt(variance / latencies_us.size());

        size_t p50_idx = (size_t)(latencies_us.size() * 0.50);
        size_t p90_idx = (size_t)(latencies_us.size() * 0.90);
        size_t p99_idx = (size_t)(latencies_us.size() * 0.99);
        size_t p999_idx = (size_t)(latencies_us.size() * 0.999);

        std::cout << "\n------------------- Latency Statistics -------------------\n";
        std::cout << "Min Latency    : " << std::fixed << std::setprecision(3) << min_lat << " us\n";
        std::cout << "Avg Latency    : " << std::fixed << std::setprecision(3) << avg_lat << " us\n";
        std::cout << "Max Latency    : " << std::fixed << std::setprecision(3) << max_lat << " us\n";
        std::cout << "Jitter (StdDev): " << std::fixed << std::setprecision(3) << jitter_stddev << " us\n";
        std::cout << "\n---------------------- Percentiles ----------------------\n";
        std::cout << "50th Percentile: " << std::fixed << std::setprecision(3) << latencies_us[p50_idx] << " us\n";
        std::cout << "90th Percentile: " << std::fixed << std::setprecision(3) << latencies_us[p90_idx] << " us\n";
        std::cout << "99th Percentile: " << std::fixed << std::setprecision(3) << latencies_us[p99_idx] << " us\n";
        std::cout << "99.9th Perc.   : " << std::fixed << std::setprecision(3) << latencies_us[p999_idx] << " us\n";
        std::cout << "==========================================================\n";
    }

    close(fd);
    return 0;
}
